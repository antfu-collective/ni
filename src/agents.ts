const npmRun = (agent: string) => (args: string[]) => {
  if (args.length > 1)
    return `${agent} run ${args[0]} -- ${args.slice(1).join(' ')}`
  else return `${agent} run ${args[0]}`
}

export const AGENTS = {
  npm: {
    'run': npmRun('npm'),
    'install': 'npm i',
    'frozen': 'npm ci',
    'global': 'npm i -g {0}',
    'add': 'npm i {0}',
    'upgrade': 'npm update {0}',
    'upgrade-interactive': null,
    'execute': 'npx {0}',
    'uninstall': 'npm uninstall {0}',
    'global_uninstall': 'npm uninstall -g {0}',
  },
  yarn: {
    'run': 'yarn run {0}',
    'install': 'yarn install',
    'frozen': 'yarn install --frozen-lockfile',
    'global': 'yarn global add {0}',
    'add': 'yarn add {0}',
    'upgrade': 'yarn upgrade {0}',
    'upgrade-interactive': 'yarn upgrade-interactive',
    'execute': 'yarn dls {0}',
    'uninstall': 'yarn remove {0}',
    'global_uninstall': 'yarn global remove {0}',
  },
  pnpm: {
    'run': npmRun('pnpm'),
    'install': 'pnpm i',
    'frozen': 'pnpm i --frozen-lockfile',
    'global': 'pnpm i -g {0}',
    'add': 'pnpm i {0}',
    'upgrade': 'pnpm update {0}',
    'upgrade-interactive': 'pnpm update -i',
    'execute': 'pnpm dlx {0}',
    'uninstall': 'pnpm remove {0}',
    'global_uninstall': 'pnpm remove -g {0}',
  },
}

export type Agent = keyof typeof AGENTS
export type Command = keyof typeof AGENTS.npm

export const agents = Object.keys(AGENTS) as Agent[]

export const LOCKS: Record<string, Agent> = {
  'pnpm-lock.yaml': 'pnpm',
  'yarn.lock': 'yarn',
  'package-lock.json': 'npm',
}

export const INSTALL_PAGE: Record<Agent, string> = {
  pnpm: 'https://pnpm.js.org/en/installation',
  yarn: 'https://yarnpkg.com/getting-started/install',
  npm: 'https://www.npmjs.com/get-npm',
}
