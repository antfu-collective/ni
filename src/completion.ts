// Print completion script
export const rawCompletionScript = `
###-begin-nr-completion-###

if type complete &>/dev/null; then
  _nr_completion() {
    local words
    local cur
    local cword
    _get_comp_words_by_ref -n =: cur words cword
    IFS=$'\\n'
    COMPREPLY=($(COMP_CWORD=$cword COMP_LINE=$cur nr --completion \${words[@]}))
  }
  complete -F _nr_completion nr
fi

###-end-nr-completion-###
`.trim()
