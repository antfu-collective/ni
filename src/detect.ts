import process from 'node:process'
import { detect as detectPM } from 'package-manager-detector'
import { x } from 'tinyexec'
import terminalLink from 'terminal-link'
import prompts from '@posva/prompts'
import { INSTALL_PAGE } from 'package-manager-detector/constants'
import { cmdExists } from './utils'

export interface DetectOptions {
  autoInstall?: boolean
  programmatic?: boolean
  cwd?: string
  /**
   * Should use Volta when present
   *
   * @see https://volta.sh/
   * @default true
   */
  detectVolta?: boolean
}

export async function detect({ autoInstall, programmatic, cwd }: DetectOptions = {}) {
  const {
    name,
    agent,
    version,
  } = await detectPM({
    cwd,
    onUnknown: (packageManager) => {
      if (!programmatic) {
        console.warn('[ni] Unknown packageManager:', packageManager)
      }
      return undefined
    },
  }) || {}

  // auto install
  if (name && !cmdExists(name) && !programmatic) {
    if (!autoInstall) {
      console.warn(`[ni] Detected ${name} but it doesn't seem to be installed.\n`)

      if (process.env.CI)
        process.exit(1)

      const link = terminalLink(name, INSTALL_PAGE[name])
      const { tryInstall } = await prompts({
        name: 'tryInstall',
        type: 'confirm',
        message: `Would you like to globally install ${link}?`,
      })
      if (!tryInstall)
        process.exit(1)
    }

    await x(
      'npm',
      ['i', '-g', `${name}${version ? `@${version}` : ''}`],
      {
        nodeOptions: {
          stdio: 'inherit',
          cwd,
        },
        throwOnError: true,
      },
    )
  }

  return agent
}
