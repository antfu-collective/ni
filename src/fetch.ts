import process from 'node:process'
import type { Choice } from '@posva/prompts'

export interface NpmPackage {
  name: string
  description: string
  version: string
  keywords: string[]
  date: string
  links: {
    npm: string
    homepage: string
    repository: string
  }
}

interface NpmRegistryResponse {
  objects: { package: NpmPackage }[]
}

export async function fetchNpmPackages(pattern: string): Promise<Choice[]> {
  const registryLink = (pattern: string) =>
        `https://registry.npmjs.com/-/v1/search?text=${pattern}&size=35`

  try {
    const result = await fetch(registryLink(pattern))
      .then(res => res.json()) as NpmRegistryResponse

    return result.objects.map(({ package: pkg }) => ({
      title: pkg.name,
      value: pkg,
      description: `${pkg.version} - ${pkg.description}`,
    }))
  }
  catch (e) {
    console.error('Error when fetching npm registry')
    process.exit(1)
  }
}
