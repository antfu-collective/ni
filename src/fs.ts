import fs from 'node:fs'
import { resolve } from 'node:path'
import type { Agent } from './agents'

const getJSONConfigFile = (cwd: string, filename: string) => {
  const path = resolve(cwd, filename)

  if (fs.existsSync(path)) {
    try {
      const raw = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(raw) as Record<string, any>
      return data
    }
    catch (e) {
      console.warn(`Failed to parse ${filename}`)
      process.exit(0)
    }
  }
}

export function getPackageJSON(cwd = process.cwd()) {
  return getJSONConfigFile(cwd, 'package.json') || {}
}

export function getDenoJSON(cwd = process.cwd()) {
  return getJSONConfigFile(cwd, 'deno.json') || getJSONConfigFile(cwd, 'deno.jsonc') || {}
}

export function getConfig(agent: Agent, cwd = process.cwd()) {
  if (agent === 'deno')
    return getDenoJSON(cwd)
  return getPackageJSON(cwd)
}
