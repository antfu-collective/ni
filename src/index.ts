export * from 'package-manager-detector/constants'
export * from 'package-manager-detector/commands'

export * from './config'
export * from './detect'
export * from './parse'
export * from './runner'
export * from './utils'
