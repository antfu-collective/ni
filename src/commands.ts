import { Agent, AGENTS, Command } from './agents'
import { exclude } from './utils'

export function getCommand(agent: Agent, commnad: Command, args: string[] = []) {
  if (!(agent in AGENTS))
    throw new Error(`Unsupported agent "${agent}"`)

  const c = AGENTS[agent][commnad]

  if (typeof c === 'function')
    return c(args)

  return c.replace('{0}', args.join(' '))
}

export function parseNi(agent: Agent, args: string[], hasLock?: boolean): string {
  if (args.length === 0)
    return getCommand(agent, 'install')

  if (args.includes('-g'))
    return getCommand(agent, 'global', exclude(args, '-g'))

  if (args.includes('--frozen-if-present')) {
    args = exclude(args, '--frozen-if-present')
    return getCommand(agent, hasLock ? 'frozen' : 'install', args)
  }

  if (args.includes('--frozen'))
    return getCommand(agent, 'frozen', exclude(args, '--frozen'))

  return getCommand(agent, 'add', args)
}

export function parseNr(agent: Agent, args: string[]): string {
  if (args.length === 0)
    args.push('start')

  return getCommand(agent, 'run', args)
}
