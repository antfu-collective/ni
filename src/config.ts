import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import ini from 'ini'
import type { Agent } from 'package-manager-detector'
import { detect } from './detect'

const customRcPath = process.env.NI_CONFIG_FILE

const home = process.platform === 'win32'
  ? process.env.USERPROFILE
  : process.env.HOME

const defaultRcPath = path.join(home || '~/', '.nirc')

const rcPath = customRcPath || defaultRcPath

interface Config {
  defaultAgent: Agent | 'prompt'
  globalAgent: Agent
}

const defaultConfig: Config = {
  defaultAgent: 'prompt',
  globalAgent: 'npm',
}

let config: Config | undefined

export async function getConfig(): Promise<Config> {
  if (!config) {
    config = Object.assign(
      {},
      defaultConfig,
      fs.existsSync(rcPath)
        ? ini.parse(fs.readFileSync(rcPath, 'utf-8'))
        : null,
    )

    if (process.env.NI_DEFAULT_AGENT)
      config.defaultAgent = process.env.NI_DEFAULT_AGENT as Agent

    if (process.env.NI_GLOBAL_AGENT)
      config.globalAgent = process.env.NI_GLOBAL_AGENT as Agent

    const agent = await detect({ programmatic: true })
    if (agent)
      config.defaultAgent = agent
  }

  return config
}

export async function getDefaultAgent(programmatic?: boolean) {
  const { defaultAgent } = await getConfig()
  if (defaultAgent === 'prompt' && (programmatic || process.env.CI))
    return 'npm'
  return defaultAgent
}

export async function getGlobalAgent() {
  const { globalAgent } = await getConfig()
  return globalAgent
}
