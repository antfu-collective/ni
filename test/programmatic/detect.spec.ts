import path from 'node:path'
import { tmpdir } from 'node:os'
import type { SpyInstance } from 'vitest'
import { afterAll, beforeAll, describe, expect, test, vi } from 'vitest'
import fs from 'fs-extra'
import { AGENTS, detect } from '../../src'

let basicLog: SpyInstance, errorLog: SpyInstance, warnLog: SpyInstance, infoLog: SpyInstance

const detectTest = (agent: string) => async () => {
  const cwd = await fs.mkdtemp(path.join(tmpdir(), 'ni-'))
  const fixture = path.join(__dirname, '..', 'fixtures', 'lockfile', agent)
  await fs.copy(fixture, cwd)

  expect(await detect({ programmatic: true, cwd })).toMatchSnapshot()
}

beforeAll(() => {
  basicLog = vi.spyOn(console, 'log')
  warnLog = vi.spyOn(console, 'warn')
  errorLog = vi.spyOn(console, 'error')
  infoLog = vi.spyOn(console, 'info')
})

afterAll(() => {
  vi.resetAllMocks()
})

const agents = [...Object.keys(AGENTS), 'unknown']
const fixtures = ['lockfile', 'packager']

// matrix testing of: fixtures x agents
fixtures.forEach(fixture => describe(fixture, () => agents.forEach((agent) => {
  test(agent, detectTest(agent))

  test('no logs', () => {
    expect(basicLog).not.toHaveBeenCalled()
    expect(warnLog).not.toHaveBeenCalled()
    expect(errorLog).not.toHaveBeenCalled()
    expect(infoLog).not.toHaveBeenCalled()
  })
})))
