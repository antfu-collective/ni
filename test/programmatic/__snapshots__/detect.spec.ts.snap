// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lockfile > bun 1`] = `"bun"`;

exports[`lockfile > npm 1`] = `"npm"`;

exports[`lockfile > pnpm 1`] = `"pnpm"`;

exports[`lockfile > pnpm@6 1`] = `"pnpm"`;

exports[`lockfile > unknown 1`] = `null`;

exports[`lockfile > yarn 1`] = `"yarn"`;

exports[`lockfile > yarn@berry 1`] = `"yarn"`;

exports[`packager > bun 1`] = `"bun"`;

exports[`packager > npm 1`] = `"npm"`;

exports[`packager > pnpm 1`] = `"pnpm"`;

exports[`packager > pnpm@6 1`] = `"pnpm"`;

exports[`packager > unknown 1`] = `null`;

exports[`packager > yarn 1`] = `"yarn"`;

exports[`packager > yarn@berry 1`] = `"yarn"`;
